File: tests/typesafety/python/result_chain_misuse.py
# expect: fails
# title: A Result spent as a number, rejected by the checker
# misuse-of: result_chain
from dataclasses import dataclass


@dataclass(frozen=True)
class Ok[T]:
    value: T


@dataclass(frozen=True)
class Err:
    reason: str


type Result[T] = Ok[T] | Err


def quote(text: str) -> Result[int]:
    return Ok(int(text) * 250) if text.isdigit() else Err(f"not a number: {text!r}")


def total_with_shipping(text: str) -> int:
    return quote(text) + 45

File: tests/typesafety/typescript/result_chain_misuse.ts
// expect: fails

type Ok<T> = { readonly kind: "ok"; readonly value: T };
type Err = { readonly kind: "err"; readonly reason: string };
type Result<T> = Ok<T> | Err;

function quote(text: string): Result<number> {
  return /^\d+$/.test(text)
    ? { kind: "ok", value: Number(text) * 250 }
    : { kind: "err", reason: `not a number: ${text}` };
}

export function totalWithShipping(text: string): number {
  return quote(text) + 45;
}
